db/schema_applier.hh

```diff
@@ -33,6 +33,68 @@ future<> merge_schema(sharded<db::system_keyspace>& sys_ks, distributed<service:
 // of feature_service and schema tables.
 future<> recalculate_schema_version(sharded<db::system_keyspace>& sys_ks, distributed<service::storage_proxy>& proxy, gms::feature_service& feat);
 
+enum class table_kind { table, view };
+
+struct table_selector {
+    bool all_in_keyspace = false; // If true, selects all existing tables in a keyspace plus what's in "tables";
+    std::unordered_map<table_kind, std::unordered_set<sstring>> tables;
+
+    table_selector& operator+=(table_selector&& o);
+    void add(table_kind t, sstring name);
+    void add(sstring name);
+};
+
+struct schema_complete_view {
+    schema_tables::schema_result keyspaces;
+    std::map<table_id, schema_mutations> tables;
+    schema_tables::schema_result types;
+    std::map<table_id, schema_mutations> views;
+    schema_tables::schema_result functions;
+    schema_tables::schema_result aggregates;
+    schema_tables::schema_result scylla_aggregates;
+};
+
+class schema_applier {
+    using keyspace_name = sstring;
+
+    sharded<service::storage_proxy>& _proxy;
+    sharded<db::system_keyspace>& _sys_ks;
+    bool _reload;
+
+    std::set<sstring> _keyspaces;
+    std::unordered_map<keyspace_name, table_selector> _affected_tables;
+    bool _has_tablet_mutations = false;
+
+    schema_complete_view _before;
+    schema_complete_view _after;
+
+    future<schema_complete_view> get_schema_complete_view();
+public:
+    schema_applier(
+            sharded<service::storage_proxy>& proxy,
+            sharded<db::system_keyspace>& sys_ks,
+            bool reload = false)
+            : _proxy(proxy), _sys_ks(sys_ks), _reload(reload) {};
+
+    // Gets called before mutations are applied,
+    // preferably no work should be done here but subsystem
+    // may do some snapshot of 'before' data.
+    future<> prepare(const std::vector<mutation>& muts);
```

Comment:
Is it clear at this point that the interface proposed here fits all group 0 use cases?

Or is it more like an experiment and subject to change?

Reply:
It's guided experiment I'd say. I inspected all the code and tried to figure out common denominator. I started with schema because it's most advanced, I think other modules will use ~3 of those interface functions.